cluster-autoscaler/simulator/drainability/rules/rules.go

```diff
@@ -36,6 +41,11 @@ type Rule interface {
 func Default() Rules {
 	return []Rule{
 		mirror.New(),
+		replicated.New(),
+		system.New(),
+		notsafetoevict.New(),
+		localstorage.New(),
```

Comment:
`localstorage` doesn't make sense to be used at all when `DeleteOptions.SkipNodesWithLocalStorage` is `false`. Can `DeleteOptions` be passed here to determine which rules are used instead?

Similarly, `system` is only useful when `SkipNodesWithSystemPods` is `true`.

Reply:
That trades a tiny amount of computation for coupling between rules. With the current implementation, the rules framework just registers known rules and doesn't have to have any understanding of rule conditions. Putting any amount of validation higher up in the drainability rules framework is a bad idea for modularity.

One option is to add an `Enabled(drainCtx, deleteOptions)` method to the rules interface that will be run when generating rules. However, I would argue the overhead isn't worth it. A single `Drainable()` function that returns early is minimal computation and keeps all logic in a single place.

Reply:
`rules.Default` already has to handle enablement flags to specific rules - it's just now the rules need to handle that state themselves, which makes each one more complex - some rules now have to handle a case in which they behave as if they didn't exist at all. It's not really about saving compute time - you're right that's negligible - but rather about keeping the rules logic focus on what they're supposed to be checking. I'd expect something else (now it is `rules.Default`) to collect all rules that should be used.

That being said, I don't have a strong opinion on this, so leaving it up to you.

Reply:
I implemented this in the latest commit. I'm still not convinced.

What I like about this implementation is that it attempts to keep DeleteOptions at the framework level and not pass them down to the rules. Unfortunately, we still need to pass the legacy custom controller option to the `replicated` rule until the option is deprecated, which nullifies this argument.

What I dislike about it is that it is adding complexity to the framework, whithout much benefit. While not having a rule is conceptually better than the `enabled` field within select rules, the field is actually well contained within a rule and uses minimal compute to exit early. I can see this evolving into more business logic living at the DefaultRules construction and that turning into spaghetti code. Simple registration and delegation for all business logic is a simpler model.

Please take a look and let me know what you think.

Another alternative is to add an Enabled() field to every rule, as previously suggested, but this is the worst of both worlds (all options still passed to the rules + more boilerplate).

Reply:
As mentioned before, I don't have a strong preference here - current version looks good to me, but adding `enabled` field within specific rules is also acceptable.